pkg/types/validation/installconfig.go

```diff
@@ -99,6 +101,115 @@ func ValidateInstallConfig(c *types.InstallConfig, openStackValidValuesFetcher o
+	if n == nil {
+		return
+	}
+	addresses = make(IPByField)
+	for _, network := range n.MachineNetwork {
+		addresses["machineNetwork"] = append(addresses["machineNetwork"], network.CIDR.IP)
+	}
+	for _, network := range n.ServiceNetwork {
+		addresses["serviceNetwork"] = append(addresses["serviceNetwork"], network.IP)
+	}
+	for _, network := range n.ClusterNetwork {
+		addresses["clusterNetwork"] = append(addresses["clusterNetwork"], network.CIDR.IP)
+	}
+	presence = make(IPAddressTypeByField)
+	for k, ips := range addresses {
+		sort.Slice(ips, func(i, j int) bool { return bytes.Compare(ips[i], ips[j]) < 0 })
+		for _, ip := range ips {
+			has := presence[k]
+			if ip.To4() != nil {
+				has.IPv4 = true
+				hasIPv4 = true
+			} else {
+				has.IPv6 = true
+				hasIPv6 = true
+			}
+			presence[k] = has
+		}
+	}
+	return
+}
+
+func ipSliceToStrings(ips []net.IP) []string {
+	var s []string
+	for _, ip := range ips {
+		s = append(s, ip.String())
+	}
+	return s
+}
+
+// validateNetworkingIPVersion checks parameters for consistency when the user
+// requests single-stack IPv6 or dual-stack modes.
+func validateNetworkingIPVersion(c *types.InstallConfig) field.ErrorList {
+	var allErrs field.ErrorList
+
+	hasIPv4, hasIPv6, presence, addresses := InferIPVersionFromInstallConfig(c.Networking)
+
+	switch {
+	case hasIPv4 && hasIPv6:
+		if c.Networking.NetworkType == "OpenShiftSDN" {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking", "networkType"), c.Networking.NetworkType, "dual-stack IPv4/IPv6 is not supported for this networking plugin"))
+		}
+
+		switch {
+		case c.Platform.BareMetal != nil:
+		case c.Platform.None != nil:
+		default:
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking"), "DualStack", "dual-stack IPv4/IPv6 is not supported for this platform, specify only one type of address"))
+		}
+		for k, v := range presence {
+			switch {
+			case k == "machineNetwork" && c.Platform.AWS != nil:
+				// AWS can default an ipv6 subnet
+			case v.IPv4 && !v.IPv6:
+				allErrs = append(allErrs, field.Invalid(field.NewPath("networking", k), strings.Join(ipSliceToStrings(addresses[k]), ", "), "dual-stack IPv4/IPv6 requires an IPv6 address in this list"))
+			case !v.IPv4 && v.IPv6:
+				allErrs = append(allErrs, field.Invalid(field.NewPath("networking", k), strings.Join(ipSliceToStrings(addresses[k]), ", "), "dual-stack IPv4/IPv6 requires an IPv4 address in this list"))
+			}
+		}
+
+	case hasIPv6:
+		if c.Networking.NetworkType == "OpenShiftSDN" {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking", "networkType"), c.Networking.NetworkType, "IPv6 is not supported for this networking plugin"))
+		}
+
+		switch {
+		case c.Platform.BareMetal != nil:
+		case c.Platform.None != nil:
+		default:
+			allErrs = append(allErrs, field.Invalid(field.NewPath("networking"), "IPv6", "IPv6 is not supported for this platform"))
+		}
+
+	case hasIPv4:
+		if len(c.Networking.ServiceNetwork) > 1 {
```

Comment:
This isn't just an IPv4 constraint. `len(c.Networking.ServiceNetwork)` must be 2 if `hasIPv4 && hasIPv6` and 1 otherwise. (ie, you must have exactly 1 network which must be IPv4 in the single-stack IPv4 case; exactly 1 network which must be IPv6 in the single-stack IPv6 case; and exactly 2 networks, one of which is IPv4 and one of which is IPv6, in the dual-stack case)

Reply:
I thought you said that we could have multiple service networks?  or was that just cluster and machine.

Reply:
Cluster networks are handled by CNO/openshift-sdn/ovn-kubernetes, and we support multiple of each family. Service networks are handled by the upstream kubernetes ServiceController, which requires exactly 1 per active address family. Machine networks are handled by installer/MCO/? and I don't know what we support there.

Reply:
machine networks are internal to installer, once we start adding ipv6 / dual stack providers each one will validate.  ok.